### ISSUE TITLE
Twisted 12.3.0pre1 is not pip installable.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6209]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#6209|
|Type|release blocker: regression|
|Created|2012-12-14 17:51:11Z|
|Branch|https://github.com/twisted/twisted/tree/setup-no-check-6209|

```
thomasina dreid:Downloads> pip install -d . --no-install -e svn+svn://svn.twistedmatrix.com/svn/Twisted/trunk#egg=Twisted
Obtaining Twisted from svn+svn://svn.twistedmatrix.com/svn/Twisted/trunk#egg=Twisted
  Exporting svn repository svn://svn.twistedmatrix.com/svn/Twisted/trunk to /Users/dreid/Downloads/src/twisted
  Running setup.py egg_info for package Twisted
    Traceback (most recent call last):
      File "<string>", line 14, in <module>
      File "/Users/dreid/Downloads/src/twisted/setup.py", line 59, in <module>
        main(sys.argv[1:])
      File "/Users/dreid/Downloads/src/twisted/setup.py", line 26, in main
        from twisted.python.dist import (
      File "./twisted/__init__.py", line 46, in <module>
        _checkRequirements()
      File "./twisted/__init__.py", line 44, in _checkRequirements
        raise ImportError(required + ".")
    ImportError: Twisted requires zope.interface 3.6.0 or later.
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):

  File "<string>", line 14, in <module>

  File "/Users/dreid/Downloads/src/twisted/setup.py", line 59, in <module>

    main(sys.argv[1:])

  File "/Users/dreid/Downloads/src/twisted/setup.py", line 26, in main

    from twisted.python.dist import (

  File "./twisted/__init__.py", line 46, in <module>

    _checkRequirements()

  File "./twisted/__init__.py", line 44, in _checkRequirements

    raise ImportError(required + ".")

ImportError: Twisted requires zope.interface 3.6.0 or later.

----------------------------------------
Command python setup.py egg_info failed with error code 1 in /Users/dreid/Downloads/src/twisted
Storing complete log in /Users/dreid/Library/Logs/pip.log
```

It looks like setup.py imports twisted.python.dist which triggers this error because setup() w/ install_requires hasn't happened yet.

<details><summary>Searchable metadata</summary>

```
trac-id__6209 6209
type__release_blocker__regression release blocker: regression
reporter__dreid dreid
priority__normal normal
milestone__Twisted_12_3 Twisted-12.3
branch__branches_setup_no_check_6209 branches/setup-no-check-6209
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1355507471000000 1355507471000000
changetime__1356013102000000 1356013102000000
version__None None
owner__therve therve
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Are you sure this is a regression?  If it is not, it doesn't belong in the Twisted-12.3 milestone.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It sounds like it _is_, introduced by [#5935](https://github.com/twisted/twisted/issues/5935).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [36581]) Branching to 'setup-no-check-6209'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I implemented the fix we talked about: adding a flag in setup.py and check it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

I think we should be clear about what exactly this branch is trying to achieve.  The existing code isn't terribly clear about what _it_ is trying to achieve, either, so let me try and clear that up:

  0. The goal of the existing `_checkRequirements` function is to display a traceback with a meaningful error to users as soon as possible when they don't have the required dependencies, rather than defer it to some inscrutable problem later.
  1. When you install Twisted by invoking `setup.py`, without any facility to automatically install dependencies, users should continue to see the traceback naming the version of zope.interface that they need.  This is really preferable to an apparently successful installation followed by inscrutable errors on `@implements` lines.
    1. You should get an error indicating you need to install it if `zope.interface` is not installed at all.
    2. You should get an error indicating you need to upgrade it if `zope.interface` is too old a version to have the features we need.
  2. When you install Twisted via some facility that _does_ have the ability to install dependencies automatically, such as `pip` or `easy_install`, Twisted should relay its dependency to the system in question via setuptools metadata, rather than reporting the error on its own; therefore it should allow the `twisted` package to import so that `setup.py` can inspect the metadata provided by `twisted.python.dist`.
    1. `zope.interface` should be automatically installed if it is not present.
    2. `zope.interface` should be automatically upgraded to a compatible version if the installed version is too old.
  3. An explicit non-goal here is trying to work intelligently with split-out subpackages, since distutils doesn't really have a way to work with those anyway, and few users want to attempt to install things piecemeal with pip.  (Plus, almost nobody legitimately has a need to, so we don't want to encourage it unnecessarily, although we do want to provide the option for when people do really need it in specialized deployments.)

This adds a (sort of) new constraint to twisted.python.dist: it may not now _or ever_ depend on zope.interface.  We should have a way to verify that this is the case.

Right now, point 2.2 is not satisfied, because we don't declare a dependency on a particular version of zope.interface, only that it's installed at all.  Tacking on a `>=3.6.0` in the dependency specifier ought to be sufficient, if I understand the way this metadata is processed.

Also, I'm not quite clear on one point here: is the presence of `setuptools` actually sufficient to detect the presence of an automated dependency resolution mechanism?  What happens if you just run regular `setup.py` when `setuptools` is available but we're not running under `easy_install` or `pip`?

(Note that this, sadly, isn't a review, it's just an attempt to triangulate on some requirements.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> When you install Twisted via some facility that does have the ability to install dependencies automatically, such as pip or easy_install, Twisted should relay its dependency to the system in question via setuptools metadata, rather than reporting the error on its own; therefore it should allow the twisted package to import so that setup.py can inspect the metadata provided by twisted.python.dist.

This sounds good an reasonable, but also note that importing the package being installed is almost an explicitly _unsupported_ feature in the setuptools world.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_9"></a>|@dreid set owner to @therve|
|-|-|

Thanks a lot for working on this therve, and thank you glyph for refining this ticket.

 1. We should include the correct version spec in the install_requires field.   I think this is probably ```zope.interface >= 3.6.0```.  I'm not sure if we care yet about handling the requirement for python 3.x.
 2. The setup.py changes don't actually work.  To be effective it'll need to be _before_ the ```twisted.python.dist``` import, right now ```_checkRequirements``` just raises before we even check if a good version of setuptools is available.

I tested the above with both ```python setup.py install```

```
(twisted)thomasina dreid:setup-no-check-6209> python setup.py install
Traceback (most recent call last):
  File "setup.py", line 60, in <module>
    main(sys.argv[1:])
  File "setup.py", line 26, in main
    from twisted.python.dist import (
  File "./twisted/__init__.py", line 53, in <module>
    _checkRequirements()
  File "./twisted/__init__.py", line 37, in _checkRequirements
    raise ImportError(required + ": no module named zope.interface.")
ImportError: Twisted requires zope.interface 3.6.0 or later: no module named zope.interface.
```

and ``` pip install . ```

```
(twisted)thomasina dreid:setup-no-check-6209> pip install .
Unpacking /Users/dreid/code/Twisted/branches/setup-no-check-6209
  Running setup.py egg_info for package from file:///Users/dreid/code/Twisted/branches/setup-no-check-6209
    Traceback (most recent call last):
      File "<string>", line 14, in <module>
      File "/var/folders/fb/bfn8yfln3_d_3bb8r1h_yhkc0000gn/T/pip-c1rbYp-build/setup.py", line 60, in <module>
        main(sys.argv[1:])
      File "/var/folders/fb/bfn8yfln3_d_3bb8r1h_yhkc0000gn/T/pip-c1rbYp-build/setup.py", line 26, in main
        from twisted.python.dist import (
      File "./twisted/__init__.py", line 53, in <module>
        _checkRequirements()
      File "./twisted/__init__.py", line 37, in _checkRequirements
        raise ImportError(required + ": no module named zope.interface.")
    ImportError: Twisted requires zope.interface 3.6.0 or later: no module named zope.interface.
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):

  File "<string>", line 14, in <module>

  File "/var/folders/fb/bfn8yfln3_d_3bb8r1h_yhkc0000gn/T/pip-c1rbYp-build/setup.py", line 60, in <module>

    main(sys.argv[1:])

  File "/var/folders/fb/bfn8yfln3_d_3bb8r1h_yhkc0000gn/T/pip-c1rbYp-build/setup.py", line 26, in main

    from twisted.python.dist import (

  File "./twisted/__init__.py", line 53, in <module>

    _checkRequirements()

  File "./twisted/__init__.py", line 37, in _checkRequirements

    raise ImportError(required + ": no module named zope.interface.")

ImportError: Twisted requires zope.interface 3.6.0 or later: no module named zope.interface.

----------------------------------------
Command python setup.py egg_info failed with error code 1 in /var/folders/fb/bfn8yfln3_d_3bb8r1h_yhkc0000gn/T/pip-c1rbYp-build
Storing complete log in /Users/dreid/.pip/pip.log
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Thanks for the review, and sorry for missing that part. It should work now.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to @therve|
|-|-|

I created a new virtualenv and installed it using:

{{
pip install -e svn+ssh://svn.twistedmatrix.com/svn/Twisted/branches/setup-no-check-6209#egg=twisted-dev
}}}

And everything seemed to work fine, inc. the suggested zope.interface 4.0.2 installation. +1 merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Just for the record, what's the status of automatic (buildbot) coverage of this change?

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_14"></a>|@dreid commented|
|-|-|

Replying to [glyph](#note_13):
> Just for the record, what's the status of automatic (buildbot) coverage of this change?

There is a branch on launchpad, and a merge proposal, linked to the original ticket about the easy_install builder:

https://bugs.launchpad.net/twisted-buildbot-configuration/+bug/1090530

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

I'll wait for the buildbot changes to happen before merging this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set status to `closed`|
|-|-|

(In [36604]) Merge setup-no-check-6209

Author: therve
Reviewers: dreid, thijs
Fixes: [#6209](https://github.com/twisted/twisted/issues/6209)

Don't check for zope requirements when setup.py is run with setuptools. It
fixes the installation with pip and easy_install when zope.interface is not
already at the correct version.
